File: inventario_visual/DAO/ProductoDAO.vb
Imports MySql.Data.MySqlClient
Imports System.Data

Namespace inventario_visual
    Public Class ProductoDAO
        Public Function ListarTodos() As List(Of Producto)
            Dim lista As New List(Of Producto)
            Using conn = ConexionDB.GetConnection()
                conn.Open()
                Dim sql As String = "SELECT p.*, c.nombre AS categoria_nombre FROM producto p LEFT JOIN categorias c ON p.categoria_id = c.id WHERE p.state = 1 ORDER BY p.id DESC"
                Using cmd As New MySqlCommand(sql, conn), reader = cmd.ExecuteReader()
                    While reader.Read()
                        lista.Add(Map(reader))
                    End While
                End Using
            End Using
            Return lista
        End Function

        Public Function Buscar(texto As String) As List(Of Producto)
            Dim lista As New List(Of Producto)
            Using conn = ConexionDB.GetConnection()
                conn.Open()
                Dim enc As String = AESUtil.Encriptar(texto)
                Dim sql As String = "SELECT p.*, c.nombre AS categoria_nombre FROM producto p LEFT JOIN categorias c ON p.categoria_id = c.id WHERE p.state = 1 AND (p.nombre LIKE @t OR p.sku LIKE @t OR p.nombre = @enc OR p.sku = @enc) ORDER BY p.id DESC"
                Using cmd As New MySqlCommand(sql, conn)
                    cmd.Parameters.AddWithValue("@t", "%" & texto & "%")
                    cmd.Parameters.AddWithValue("@enc", enc)
                    Using reader = cmd.ExecuteReader()
                        While reader.Read()
                            lista.Add(Map(reader))
                        End While
                    End Using
                End Using
            End Using
            Return lista
        End Function

        Public Function ObtenerPorId(id As Long) As Producto
            Using conn = ConexionDB.GetConnection()
                conn.Open()
                Dim sql As String = "SELECT p.*, c.nombre AS categoria_nombre FROM producto p LEFT JOIN categorias c ON p.categoria_id = c.id WHERE p.id = @id"
                Using cmd As New MySqlCommand(sql, conn)
                    cmd.Parameters.AddWithValue("@id", id)
                    Using reader = cmd.ExecuteReader()
                        If reader.Read() Then Return Map(reader)
                    End Using
                End Using
            End Using
            Return Nothing
        End Function

        Public Sub Insertar(p As Producto)
            Using conn = ConexionDB.GetConnection()
                conn.Open()
                Dim sql As String = "INSERT INTO producto (categoria_id, sku, nombre, descripcion, precio_venta, costo_promedio, stock_minimo, stock_actual, state, imagen) VALUES (@cat, @sku, @nombre, @desc, @pv, @pc, @min, @stock, @state, @img)"
                Using cmd As New MySqlCommand(sql, conn)
                    AddParams(cmd, p)
                    cmd.ExecuteNonQuery()
                End Using
            End Using
        End Sub

        Public Sub Actualizar(p As Producto)
            Using conn = ConexionDB.GetConnection()
                conn.Open()
                Dim sql As String = "UPDATE producto SET categoria_id=@cat, sku=@sku, nombre=@nombre, descripcion=@desc, precio_venta=@pv, costo_promedio=@pc, stock_minimo=@min, stock_actual=@stock, state=@state, imagen=@img WHERE id=@id"
                Using cmd As New MySqlCommand(sql, conn)
                    AddParams(cmd, p)
                    cmd.Parameters.AddWithValue("@id", p.Id)
                    cmd.ExecuteNonQuery()
                End Using
            End Using
        End Sub

        Public Sub Eliminar(id As Long)
            Using conn = ConexionDB.GetConnection()
                conn.Open()
                Using cmd As New MySqlCommand("DELETE FROM producto WHERE id = @id", conn)
                    cmd.Parameters.AddWithValue("@id", id)
                    cmd.ExecuteNonQuery()
                End Using
            End Using
        End Sub

        Public Function ContarTotal() As Integer
            Using conn = ConexionDB.GetConnection()
                conn.Open()
                Using cmd As New MySqlCommand("SELECT COUNT(*) FROM producto WHERE state = 1", conn)
                    Return Convert.ToInt32(cmd.ExecuteScalar())
                End Using
            End Using
        End Function

        Public Function ListarStockBajo() As List(Of Producto)
            Dim lista As New List(Of Producto)
            Using conn = ConexionDB.GetConnection()
                conn.Open()
                Dim sql As String = "SELECT p.*, c.nombre AS categoria_nombre FROM producto p LEFT JOIN categorias c ON p.categoria_id = c.id WHERE p.state = 1"
                Using cmd As New MySqlCommand(sql, conn), reader = cmd.ExecuteReader()
                    While reader.Read()
                        Dim p = Map(reader)
                        If p.StockActual <= p.StockMinimo Then lista.Add(p)
                    End While
                End Using
            End Using
            Return lista
        End Function

        Public Function ContarStockBajo() As Integer
            Return ListarStockBajo().Count
        End Function

        Public Function ListarTopProductosPorStock(limite As Integer) As List(Of Producto)
            Dim lista As New List(Of Producto)
            Using conn = ConexionDB.GetConnection()
                conn.Open()
                Dim sql As String = "SELECT p.*, c.nombre AS categoria_nombre FROM producto p LEFT JOIN categorias c ON p.categoria_id = c.id WHERE p.state = 1 ORDER BY p.stock_actual DESC LIMIT @lim"
                Using cmd As New MySqlCommand(sql, conn)
                    cmd.Parameters.AddWithValue("@lim", limite)
                    Using reader = cmd.ExecuteReader()
                        While reader.Read()
                            lista.Add(Map(reader))
                        End While
                    End Using
                End Using
            End Using
            Return lista
        End Function

        Private Shared Sub AddParams(cmd As MySqlCommand, p As Producto)
            cmd.Parameters.AddWithValue("@sku", AESUtil.Encriptar(p.Sku))
            cmd.Parameters.AddWithValue("@nombre", AESUtil.Encriptar(p.Nombre))
            cmd.Parameters.AddWithValue("@desc", AESUtil.Encriptar(p.Descripcion))
            cmd.Parameters.AddWithValue("@cat", If(p.CategoriaId.HasValue, CObj(p.CategoriaId.Value), DBNull.Value))
            cmd.Parameters.AddWithValue("@pv", p.PrecioVenta)
            cmd.Parameters.AddWithValue("@pc", p.PrecioCompra)
            cmd.Parameters.AddWithValue("@stock", p.StockActual)
            cmd.Parameters.AddWithValue("@min", p.StockMinimo)
            cmd.Parameters.AddWithValue("@state", If(p.State, 1, 0))
            cmd.Parameters.AddWithValue("@img", If(String.IsNullOrEmpty(p.Imagen), CObj(DBNull.Value), p.Imagen))
        End Sub

        Private Shared Function Map(reader As MySqlDataReader) As Producto
            Return New Producto With {
                .Id = Convert.ToInt64(reader("id")),
                .Sku = AESUtil.Desencriptar(reader("sku").ToString()),
                .Nombre = AESUtil.Desencriptar(reader("nombre").ToString()),
                .Descripcion = AESUtil.Desencriptar(If(IsDBNull(reader("descripcion")), "", reader("descripcion").ToString())),
                .CategoriaId = If(IsDBNull(reader("categoria_id")), CType(Nothing, Long?), Convert.ToInt64(reader("categoria_id"))),
                .PrecioVenta = Convert.ToDouble(reader("precio_venta")),
                .PrecioCompra = Convert.ToDouble(reader("costo_promedio")),
                .StockActual = Convert.ToDouble(reader("stock_actual")),
                .StockMinimo = Convert.ToDouble(reader("stock_minimo")),
                .State = Convert.ToBoolean(reader("state")),
                .Imagen = If(IsDBNull(reader("imagen")), "", reader("imagen").ToString()),
                .CategoriaNombre = AESUtil.Desencriptar(If(IsDBNull(reader("categoria_nombre")), "", reader("categoria_nombre").ToString()))
            }
        End Function
    End Class
End Namespace

File: inventario_python/dao/producto_dao.py
from db.conexion import conectar
from models.producto import Producto
from utils.aes_util import encriptar, desencriptar


class ProductoDAO:

    def _mapear(self, row):
        p = Producto()
        p.id = row[0]
        p.categoria_id = row[1]
        p.sku = desencriptar(row[2]) if row[2] else ""
        p.nombre = desencriptar(row[3]) if row[3] else ""
        p.descripcion = desencriptar(row[4]) if row[4] else ""
        p.precio_venta = row[5] or 0.0
        p.costo_promedio = row[6] or 0.0
        p.stock_minimo = row[7] or 0.0
        p.stock_actual = row[8] or 0.0
        p.state = bool(row[9])
        p.imagen = row[10]
        if len(row) > 11 and row[11]:
            p.categoria_nombre = desencriptar(row[11])
        return p

    def listar_todos(self):
        conn = conectar()
        if not conn:
            return []
        try:
            cur = conn.cursor()
            cur.execute("SELECT * FROM producto ORDER BY id DESC")
            rows = cur.fetchall()
            return [self._mapear(r) for r in rows]
        finally:
            conn.close()

    def listar_todos_con_categoria(self):
        conn = conectar()
        if not conn:
            return []
        try:
            cur = conn.cursor()
            cur.execute(
                "SELECT p.*, c.nombre as categoria_nombre FROM producto p "
                "LEFT JOIN categorias c ON p.categoria_id = c.id ORDER BY p.id DESC"
            )
            rows = cur.fetchall()
            return [self._mapear(r) for r in rows]
        finally:
            conn.close()

    def buscar(self, texto):
        conn = conectar()
        if not conn:
            return []
        try:
            cur = conn.cursor()
            enc = encriptar(texto)
            cur.execute(
                "SELECT * FROM producto WHERE nombre LIKE %s OR sku LIKE %s OR nombre = %s OR sku = %s ORDER BY id DESC",
                (f"%{texto}%", f"%{texto}%", enc, enc),
            )
            rows = cur.fetchall()
            return [self._mapear(r) for r in rows]
        finally:
            conn.close()

    def listar_stock_bajo(self):
        conn = conectar()
        if not conn:
            return []
        try:
            cur = conn.cursor()
            cur.execute("SELECT * FROM producto WHERE stock_actual <= stock_minimo AND state = 1 ORDER BY id DESC")
            rows = cur.fetchall()
            return [self._mapear(r) for r in rows]
        finally:
            conn.close()

    def obtener_por_id(self, id):
        conn = conectar()
        if not conn:
            return None
        try:
            cur = conn.cursor()
            cur.execute("SELECT * FROM producto WHERE id = %s", (id,))
            row = cur.fetchone()
            return self._mapear(row) if row else None
        finally:
            conn.close()

    def insertar(self, p):
        conn = conectar()
        if not conn:
            return False
        try:
            cur = conn.cursor()
            cur.execute(
                "INSERT INTO producto (categoria_id, sku, nombre, descripcion, precio_venta, costo_promedio, stock_minimo, stock_actual, state, imagen) "
                "VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s,%s)",
                (p.categoria_id, encriptar(p.sku) if p.sku else None, encriptar(p.nombre),
                 encriptar(p.descripcion) if p.descripcion else None,
                 p.precio_venta, p.costo_promedio, p.stock_minimo, p.stock_actual, 1 if p.state else 0, p.imagen),
            )
            conn.commit()
            return True
        except Exception as e:
            conn.rollback()
            raise e
        finally:
            conn.close()

    def actualizar(self, p):
        conn = conectar()
        if not conn:
            return False
        try:
            cur = conn.cursor()
            cur.execute(
                "UPDATE producto SET categoria_id=%s, sku=%s, nombre=%s, descripcion=%s, precio_venta=%s, "
                "costo_promedio=%s, stock_minimo=%s, state=%s, imagen=%s WHERE id=%s",
                (p.categoria_id, encriptar(p.sku) if p.sku else None, encriptar(p.nombre),
                 encriptar(p.descripcion) if p.descripcion else None,
                 p.precio_venta, p.costo_promedio, p.stock_minimo, 1 if p.state else 0, p.imagen, p.id),
            )
            conn.commit()
            return True
        except Exception as e:
            conn.rollback()
            raise e
        finally:
            conn.close()

    def eliminar(self, id):
        conn = conectar()
        if not conn:
            return False
        try:
            cur = conn.cursor()
            cur.execute("DELETE FROM producto WHERE id = %s", (id,))
            conn.commit()
            return True
        except Exception as e:
            conn.rollback()
            raise e
        finally:
            conn.close()

    def contar_total(self):
        conn = conectar()
        if not conn:
            return 0
        try:
            cur = conn.cursor()
            cur.execute("SELECT COUNT(*) FROM producto")
            return cur.fetchone()[0]
        finally:
            conn.close()

    def contar_stock_bajo(self):
        conn = conectar()
        if not conn:
            return 0
        try:
            cur = conn.cursor()
            cur.execute("SELECT COUNT(*) FROM producto WHERE stock_actual <= stock_minimo AND state = 1")
            return cur.fetchone()[0]
        finally:
            conn.close()

    def listar_top_productos_por_stock(self, limite=10):
        conn = conectar()
        if not conn:
            return []
        try:
            cur = conn.cursor()
            cur.execute("SELECT * FROM producto ORDER BY stock_actual DESC LIMIT %s", (limite,))
            return [self._mapear(r) for r in cur.fetchall()]
        finally:
            conn.close()
